// @adeunis/codecs/src/products/comfort_v2/0x4c.parser.ts
namespace codec {
  /**
   * Comfort 2 0x4c (historic data) frame parser
   */
  export class Comfort20x4cParser implements FrameParser {
    readonly deviceType = 'comfort2';
    readonly frameCode = 0x4c;

    public parseFrame(payload: Buffer, configuration: Buffer, network: Network) {
      const appContent: Content = { type: '0x4c Comfort 2 data' };
      const payloadLength = payload[1] & 0x04 ? payload.length - 4 : payload.length;

      if (payload[1] & 0x04) {
        // retrieval of the frame date timestamp value, then conversion to epoch 2013 (+1356998400), then in second
        const myDate = new Date((payload.readUInt32BE(payloadLength) + 1356998400) * 1000);
        // As we do not know if timestamp is locale or UTC , do remove 'UTC' indication
        appContent['timestamp'] = myDate.toJSON().replace('Z', '');
      }
      let rawValue;
      const temp = [],
        humidity = [];

      // Loop through historic data [t=0, t-1, t-2,...]
      for (let offset = 2; offset < payloadLength; offset += 3) {
        rawValue = payload.readInt16BE(offset);
        temp.push(rawValue / 10);

        rawValue = payload.readUInt8(offset + 2);
        humidity.push(rawValue);
      }

      appContent['decodingInfo'] = 'values: [t=0, t-1, t-2, ...]';
      appContent['temperature'] = { unit: '\u00B0' + 'C', values: temp };
      appContent['humidity'] = { unit: '\u0025', values: humidity };

      return appContent;
    }
  }
}
