// @adeunis/codecs/dist/ttn/pulse_lib.js
"use strict";
Object.prototype.readInt8 = function (offset) {
    var buffer = this;
    var a = buffer[offset];
    if ((a & 0x80) > 0) {
        return a - 0x100;
    }
    return a;
};
Object.prototype.readUInt8 = function (offset) {
    var buffer = this;
    return buffer[offset];
};
Object.prototype.readInt16BE = function (offset) {
    var buffer = this;
    var a = (buffer[offset] << 8) | buffer[offset + 1];
    if ((a & 0x8000) > 0) {
        return a - 0x10000;
    }
    return a;
};
Object.prototype.readUInt16BE = function (offset) {
    var buffer = this;
    return (buffer[offset] << 8) | buffer[offset + 1];
};
Object.prototype.readInt32BE = function (offset) {
    var buffer = this;
    var a = ((buffer[offset] << 24) | (buffer[offset + 1] << 16) | (buffer[offset + 2] << 8) | buffer[offset + 3]) >>> 0;
    if ((a & 0x80000000) > 0) {
        return a - 0x10000000;
    }
    return a;
};
Object.prototype.readUInt32BE = function (offset) {
    var buffer = this;
    return ((buffer[offset] << 24) | (buffer[offset + 1] << 16) | (buffer[offset + 2] << 8) | buffer[offset + 3]) >>> 0;
};
Object.prototype.readFloatBE = function (offset) {
    var buffer = this;
    var value = ((buffer[offset] << 24) | (buffer[offset + 1] << 16) | (buffer[offset + 2] << 8) | buffer[offset + 3]) >>> 0;
    return new Float32Array(new Uint32Array([value]).buffer)[0];
};
if (typeof module !== 'undefined') {
    module.exports = codec;
}
if (typeof process !== 'undefined' && process.env['NODE_ENV'] === 'test') {
    global.codec = codec;
}
var codec;
(function (codec) {
    var PartialDecodingReason;
    (function (PartialDecodingReason) {
        PartialDecodingReason[PartialDecodingReason["NONE"] = 0] = "NONE";
        PartialDecodingReason[PartialDecodingReason["MISSING_NETWORK"] = 1] = "MISSING_NETWORK";
        PartialDecodingReason[PartialDecodingReason["MISSING_CONFIGURATION"] = 2] = "MISSING_CONFIGURATION";
    })(PartialDecodingReason = codec.PartialDecodingReason || (codec.PartialDecodingReason = {}));
})(codec || (codec = {}));
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var codec;
(function (codec) {
    var ContentImpl = (function () {
        function ContentImpl(type) {
            if (type === void 0) { type = 'Unknown'; }
            this.type = type;
            this.partialDecoding = codec.PartialDecodingReason.NONE;
        }
        ContentImpl.merge = function () {
            var contents = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                contents[_i] = arguments[_i];
            }
            if (!contents || contents.length === 0) {
                return null;
            }
            return Object.assign.apply(Object, __spreadArray([new ContentImpl(contents[0].type)], contents, false));
        };
        ContentImpl.prototype.merge = function () {
            var contents = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                contents[_i] = arguments[_i];
            }
            return ContentImpl.merge.apply(ContentImpl, __spreadArray([this], contents, false));
        };
        return ContentImpl;
    }());
    codec.ContentImpl = ContentImpl;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var PlateformCommonUtils = (function () {
        function PlateformCommonUtils() {
        }
        PlateformCommonUtils.getProductModeText = function (value) {
            switch (value) {
                case 0:
                    return 'PARK';
                case 1:
                    return 'PRODUCTION';
                case 2:
                    return 'TEST';
                case 3:
                    return 'DEAD';
                default:
                    return '';
            }
        };
        return PlateformCommonUtils;
    }());
    codec.PlateformCommonUtils = PlateformCommonUtils;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var GenericStatusByteParser = (function () {
        function GenericStatusByteParser() {
            this.deviceType = 'any';
            this.frameCode = 0;
        }
        GenericStatusByteParser.prototype.parseFrame = function (payload, configuration) {
            var statusContent = {};
            statusContent['frameCounter'] = (payload[1] & 0xe0) >> 5;
            statusContent['hardwareError'] = false;
            statusContent['lowBattery'] = Boolean(payload[1] & 0x02);
            statusContent['configurationDone'] = Boolean(payload[1] & 0x01);
            return statusContent;
        };
        return GenericStatusByteParser;
    }());
    codec.GenericStatusByteParser = GenericStatusByteParser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Generic0x20Parser = (function () {
        function Generic0x20Parser() {
            this.deviceType = 'any';
            this.frameCode = 0x20;
        }
        Generic0x20Parser.prototype.parseFrame = function (payload, configuration, network, deviceType) {
            var appContent = { type: '0x20 Configuration' };
            switch (payload.length) {
                case 4:
                    appContent['loraAdr'] = Boolean(payload[2] & 0x01);
                    appContent['loraProvisioningMode'] = payload[3] === 0 ? 'ABP' : 'OTAA';
                    if (deviceType !== 'analog' && deviceType !== 'drycontacts' && deviceType !== 'pulse' && deviceType !== 'temp') {
                        appContent['loraDutycyle'] = payload[2] & 0x04 ? 'activated' : 'deactivated';
                        appContent['loraClassMode'] = payload[2] & 0x20 ? 'CLASS C' : 'CLASS A';
                    }
                    break;
                case 3:
                case 5:
                    appContent['sigfoxRetry'] = payload[2] & 0x03;
                    if (payload.length === 5) {
                        appContent['sigfoxDownlinkPeriod'] = { unit: 'm', value: payload.readInt16BE(3) };
                    }
                    break;
                default:
                    appContent.partialDecoding = codec.PartialDecodingReason.MISSING_NETWORK;
                    break;
            }
            return appContent;
        };
        return Generic0x20Parser;
    }());
    codec.Generic0x20Parser = Generic0x20Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x10Parser = (function () {
        function Pulse0x10Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x10;
            this.pulse0x11Parser = new codec.Pulse0x11Parser();
            this.pulse0x12Parser = new codec.Pulse0x12Parser();
        }
        Pulse0x10Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x10 Pulse configuration' };
            var chA = { name: 'channel A' };
            var chB = { name: 'channel B' };
            appContent['productMode'] = codec.PlateformCommonUtils.getProductModeText(payload[2]);
            var knownNetwork = this.inferNetwork(payload.length);
            var offset = 0;
            if (payload[8] === 2) {
                if (knownNetwork === 'sigfox') {
                    appContent['transmissionPeriod'] = { unit: 's', value: payload[3] * 20 };
                    offset = -1;
                }
                else {
                    appContent['transmissionPeriod'] = { unit: 's', value: payload.readUInt16BE(3) * 20 };
                }
            }
            else {
                if (knownNetwork === 'sigfox') {
                    appContent['transmissionPeriod'] = { unit: 'm', value: payload[3] * 10 };
                    offset = -1;
                }
                else {
                    appContent['transmissionPeriod'] = { unit: 'm', value: payload.readUInt16BE(3) };
                }
            }
            if (payload[2] === 2) {
                appContent['flowCalculationPeriod'] = { unit: 's', value: payload.readUInt16BE(offset + 8) * 20 };
            }
            else {
                appContent['flowCalculationPeriod'] = { unit: 'm', value: payload.readUInt16BE(offset + 8) };
            }
            appContent['historicMode'] = this.getHistoricModeText(payload[offset + 6]);
            chA['state'] = this.getStateText(Boolean(payload[offset + 5] & 0x01));
            chA['type'] = this.getTypeText(Boolean(payload[offset + 5] & 0x02));
            chA['tamperActivated'] = Boolean(payload[offset + 5] & 0x08);
            chA['debouncingPeriod'] = {
                unit: 'ms',
                value: this.getDebouncingPeriod(payload[offset + 7] & 0x0f),
            };
            if (knownNetwork === 'lora868') {
                chA['leakageDetection'] = {
                    overflowAlarmTriggerThreshold: payload.readUInt16BE(10),
                    threshold: payload.readUInt16BE(14),
                    dailyPeriodsBelowWhichLeakageAlarmTriggered: payload.readUInt16BE(18),
                };
            }
            chB['state'] = this.getStateText(Boolean(payload[offset + 5] & 0x10));
            chB['type'] = this.getTypeText(Boolean(payload[offset + 5] & 0x20));
            chB['tamperActivated'] = Boolean(payload[offset + 5] & 0x80);
            chB['debouncingPeriod'] = {
                unit: 'ms',
                value: this.getDebouncingPeriod((payload[offset + 7] & 0xf0) >> 4),
            };
            if (knownNetwork === 'lora868') {
                chB['leakageDetection'] = {
                    overflowAlarmTriggerThreshold: payload.readUInt16BE(12),
                    threshold: payload.readUInt16BE(16),
                    dailyPeriodsBelowWhichLeakageAlarmTriggered: payload.readUInt16BE(20),
                };
            }
            appContent['channels'] = [chA, chB];
            return appContent;
        };
        Pulse0x10Parser.prototype.inferNetwork = function (length) {
            switch (length) {
                case 22:
                    return 'lora868';
                case 9:
                    return 'sigfox';
                default:
                    return 'unknown';
            }
        };
        Pulse0x10Parser.prototype.getStateText = function (value) {
            return value ? 'enabled' : 'disabled';
        };
        Pulse0x10Parser.prototype.getTypeText = function (value) {
            return value ? 'gasPullUpOn' : 'otherPullUpOff';
        };
        Pulse0x10Parser.prototype.getHistoricModeText = function (value) {
            switch (value) {
                case 0:
                    return 'noHistoric';
                case 1:
                    return 'historicLogEvery10min';
                case 2:
                    return 'historicLogEvery1h';
                default:
                    return '';
            }
        };
        Pulse0x10Parser.prototype.getDebouncingPeriod = function (value) {
            switch (value) {
                case 0:
                    return 0;
                case 1:
                    return 1;
                case 2:
                    return 10;
                case 3:
                    return 20;
                case 4:
                    return 50;
                case 5:
                    return 100;
                case 6:
                    return 200;
                case 7:
                    return 500;
                case 8:
                    return 1000;
                case 9:
                    return 2000;
                case 10:
                    return 5000;
                case 11:
                    return 10000;
                default:
                    return 0;
            }
        };
        return Pulse0x10Parser;
    }());
    codec.Pulse0x10Parser = Pulse0x10Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x11Parser = (function () {
        function Pulse0x11Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x11;
        }
        Pulse0x11Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x11 Pulse configuration' };
            var chA = { name: 'channel A' };
            var chB = { name: 'channel B' };
            chA['leakageDetection'] = {
                overflowAlarmTriggerThreshold: payload.readUInt16BE(2),
                threshold: payload.readUInt16BE(6),
            };
            chB['leakageDetection'] = {
                overflowAlarmTriggerThreshold: payload.readUInt16BE(4),
                threshold: payload.readUInt16BE(8),
            };
            appContent['channels'] = [chA, chB];
            return appContent;
        };
        return Pulse0x11Parser;
    }());
    codec.Pulse0x11Parser = Pulse0x11Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x12Parser = (function () {
        function Pulse0x12Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x12;
        }
        Pulse0x12Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x12 Pulse configuration' };
            var chA = { name: 'channel A' };
            var chB = { name: 'channel B' };
            chA['leakageDetection'] = {
                dailyPeriodsBelowWhichLeakageAlarmTriggered: payload.readUInt16BE(2),
            };
            chB['leakageDetection'] = {
                dailyPeriodsBelowWhichLeakageAlarmTriggered: payload.readUInt16BE(4),
            };
            appContent['channels'] = [chA, chB];
            return appContent;
        };
        return Pulse0x12Parser;
    }());
    codec.Pulse0x12Parser = Pulse0x12Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x30Parser = (function () {
        function Pulse0x30Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x30;
        }
        Pulse0x30Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x30 Pulse keep alive' };
            var chA = { name: 'channel A' };
            var chB = { name: 'channel B' };
            chA['flow'] = {
                alarm: Boolean(payload[2] & 0x01),
                last24hMin: payload.readUInt16BE(7),
                last24hMax: payload.readUInt16BE(3),
            };
            chA['tamperAlarm'] = Boolean(payload[2] & 0x04);
            chA['leakageAlarm'] = Boolean(payload[2] & 0x10);
            chB['flow'] = {
                alarm: Boolean(payload[2] & 0x02),
                last24hMin: payload.readUInt16BE(9),
                last24hMax: payload.readUInt16BE(5),
            };
            chB['tamperAlarm'] = Boolean(payload[2] & 0x08);
            chB['leakageAlarm'] = Boolean(payload[2] & 0x20);
            appContent['channels'] = [chA, chB];
            return appContent;
        };
        return Pulse0x30Parser;
    }());
    codec.Pulse0x30Parser = Pulse0x30Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x46Parser = (function () {
        function Pulse0x46Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x46;
        }
        Pulse0x46Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x46 Pulse data' };
            appContent['decodingInfo'] = 'counterValues: [Channel A, Channel B]';
            appContent['counterValues'] = [payload.readUInt32BE(2), payload.readUInt32BE(6)];
            return appContent;
        };
        return Pulse0x46Parser;
    }());
    codec.Pulse0x46Parser = Pulse0x46Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x47Parser = (function () {
        function Pulse0x47Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x47;
        }
        Pulse0x47Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x47 Pulse alarm' };
            appContent['decodingInfo'] = 'flowValues: [Channel A, Channel B]';
            appContent['flowValues'] = [payload.readUInt16BE(2), payload.readUInt16BE(4)];
            return appContent;
        };
        return Pulse0x47Parser;
    }());
    codec.Pulse0x47Parser = Pulse0x47Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var Pulse0x48Parser = (function () {
        function Pulse0x48Parser() {
            this.deviceType = 'pulse';
            this.frameCode = 0x48;
        }
        Pulse0x48Parser.prototype.parseFrame = function (payload, configuration, network) {
            var appContent = { type: '0x48 Pulse historic data' };
            var deltasA = [];
            var deltasB = [];
            var chA = { name: 'channel A' };
            var chB = { name: 'channel B' };
            var frameIndex = payload[2];
            appContent['frameIndex'] = frameIndex;
            if (frameIndex === 0) {
                chA['index'] = payload.readUInt32BE(3);
                chB['index'] = payload.readUInt32BE(7);
            }
            var start = frameIndex === 0 ? 11 : 3;
            var base = this.getBase(frameIndex, payload.length);
            for (var offset = start; offset < payload.length; offset += 4) {
                deltasA.push(payload.readUInt16BE(offset));
                deltasB.push(payload.readUInt16BE(offset + 2));
            }
            if (deltasA.length > 0) {
                appContent['baseTime'] = base;
                appContent['decodingInfo'] = "deltaValues: [t=".concat(base, " to t=").concat(base + 1, ", t=").concat(base + 1, " to t=").concat(base + 2, ", ...]");
                chA['deltaValues'] = deltasA;
                chB['deltaValues'] = deltasB;
            }
            appContent['channels'] = [chA, chB];
            return appContent;
        };
        Pulse0x48Parser.prototype.getBase = function (frameIndex, byteLength) {
            if (byteLength === 31 || byteLength === 51) {
                return [1, 11][frameIndex];
            }
            else if ((frameIndex === 2 && byteLength === 7) || (frameIndex === 12 && byteLength === 7)) {
                return 23;
            }
            else if (frameIndex === 3 && byteLength === 7) {
                return 5;
            }
            else if (byteLength === 11) {
                return [1, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23][frameIndex];
            }
            else {
                return 0;
            }
        };
        return Pulse0x48Parser;
    }());
    codec.Pulse0x48Parser = Pulse0x48Parser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var PulseStatusByteParser = (function () {
        function PulseStatusByteParser() {
            this.deviceType = 'pulse';
            this.frameCode = 0;
        }
        PulseStatusByteParser.prototype.parseFrame = function (payload, configuration) {
            var statusContent = {};
            var parser = new codec.GenericStatusByteParser();
            statusContent = parser.parseFrame(payload, configuration);
            return { status: statusContent };
        };
        return PulseStatusByteParser;
    }());
    codec.PulseStatusByteParser = PulseStatusByteParser;
})(codec || (codec = {}));
var codec;
(function (codec) {
    var CommonDecoder = (function () {
        function CommonDecoder() {
        }
        CommonDecoder.prototype.decode = function (payload) {
            var _this = this;
            var frameCode = payload[0];
            var configuration;
            if (frameCode === 0x10) {
                configuration = payload;
            }
            var activeParsers = this.getActiveParsers(frameCode);
            var partialContents = activeParsers.map(function (p) {
                var partialContent;
                try {
                    partialContent = p.parseFrame(payload, configuration, 'unknown', _this.deviceType);
                }
                catch (error) {
                    partialContent = { error: error.toString() };
                }
                return partialContent;
            });
            if (activeParsers.every(function (p) { return p.frameCode < 0; })) {
                partialContents.push({ type: 'Unsupported' });
            }
            var content = Object.assign.apply(Object, __spreadArray([{}], partialContents, false));
            var typestr = content['type'];
            delete content['type'];
            content = Object.assign({ type: typestr }, content);
            return content;
        };
        CommonDecoder.prototype.getActiveParsers = function (frameCode) {
            var activeParsers = [];
            return activeParsers;
        };
        return CommonDecoder;
    }());
    codec.CommonDecoder = CommonDecoder;
})(codec || (codec = {}));
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var codec;
(function (codec) {
    var Decoder = (function (_super) {
        __extends(Decoder, _super);
        function Decoder() {
            var _this = _super.call(this) || this;
            _this.deviceType = 'pulse';
            return _this;
        }
        Decoder.prototype.getActiveParsers = function (frameCode) {
            var activeParsers = [];
            var statusByteParsers = new codec.PulseStatusByteParser();
            var dataParser;
            switch (frameCode) {
                case 0x10:
                    dataParser = new codec.Pulse0x10Parser();
                    break;
                case 0x11:
                    dataParser = new codec.Pulse0x11Parser();
                    break;
                case 0x12:
                    dataParser = new codec.Pulse0x12Parser();
                    break;
                case 0x30:
                    dataParser = new codec.Pulse0x30Parser();
                    break;
                case 0x46:
                    dataParser = new codec.Pulse0x46Parser();
                    break;
                case 0x47:
                    dataParser = new codec.Pulse0x47Parser();
                    break;
                case 0x48:
                    dataParser = new codec.Pulse0x48Parser();
                    break;
                case 0x20:
                    dataParser = new codec.Generic0x20Parser();
                    break;
                default:
                    return activeParsers;
            }
            activeParsers = activeParsers.concat(statusByteParsers);
            activeParsers = activeParsers.concat(dataParser);
            return activeParsers;
        };
        return Decoder;
    }(codec.CommonDecoder));
    codec.Decoder = Decoder;
})(codec || (codec = {}));
function decodeUplink(input) {
    var decoder = new codec.Decoder();
    return {
        data: {
            bytes: decoder.decode(input.bytes),
        },
        warnings: [],
        errors: [],
    };
}
