// @adeunis/codecs/src/products/breath/0x6e.parser.ts
namespace codec {
  /**
   * Breath 0x6e (alarm) frame parser
   */
  export class Breath0x6eParser implements FrameParser {
    readonly deviceType = 'breath';
    readonly frameCode = 0x6e;

    public parseFrame(payload: Buffer, configuration: Buffer, network: Network) {
      const appContent: Content = { type: '0x6e Breath alarm' };

      appContent['tvoc'] = {
        alarmStatus: payload[2] & 0x01 ? 'active' : 'inactive',
        unit: 'µg/m3',
        value: payload.readUInt16BE(3),
      };
      appContent['pm10'] = {
        alarmStatus: payload[2] & 0x02 ? 'active' : 'inactive',
        unit: 'µg/m3',
        value: payload.readUInt16BE(5),
      };
      appContent['pm25'] = {
        alarmStatus: payload[2] & 0x04 ? 'active' : 'inactive',
        unit: 'µg/m3',
        value: payload.readUInt16BE(7),
      };
      appContent['pm1'] = {
        alarmStatus: payload[2] & 0x08 ? 'active' : 'inactive',
        unit: 'µg/m3',
        value: payload.readUInt16BE(9),
      };

      return appContent;
    }
  }
}
