// @adeunis/codecs/src/products/comfort_serenity/0x31.parser.ts
namespace codec {
  /**
   * Comfort Serenity 0x31 daily frame parser
   */
  export class ComfortSerenity0x31Parser implements FrameParser {
    readonly deviceType = 'comfortSerenity';
    readonly frameCode = 0x31;

    public parseFrame(payload: Buffer, configuration: Buffer, network: Network) {
      const appContent: Content = { type: '0x31 Comfort Serenity Get Register' };
      const payloadLength = payload[1] & 0x04 ? payload.length - 4 : payload.length;

      if (payload[1] & 0x04) {
        // retrieval of the frame date timestamp value, then conversion to epoch 2013 (+1356998400), then in second
        const myDate = new Date((payload.readUInt32BE(payloadLength) + 1356998400) * 1000);
        // As we do not know if timestamp is locale or UTC , do remove 'UTC' indication
        appContent['timestamp'] = myDate.toJSON().replace('Z', '');
      }

      appContent['value1'] = payload.readUInt16BE(2);
      appContent['value2'] = payload.readUInt8(4);
      appContent['value3'] = payload.readUInt32BE(5);

      return appContent;
    }
  }
}
