// @adeunis/codecs/src/products/comfort_serenity/0x38.parser.ts
namespace codec {
  /**
   * Comfort Serenity 0x38 daily frame parser
   */
  export class ComfortSerenity0x38Parser implements FrameParser {
    readonly deviceType = 'comfortSerenity';
    readonly frameCode = 0x38;

    public parseFrame(payload: Buffer, configuration: Buffer, network: Network) {
      const appContent: Content = { type: '0x38 Comfort Serenity Daily frame' };

      if (payload[1] & 0x04) {
        // retrieval of the frame date timestamp value, then conversion to epoch 2013 (+1356998400), then in second
        const myDate = new Date((payload.readUInt32BE(25) + 1356998400) * 1000);
        // As we do not know if timestamp is locale or UTC , do remove 'UTC' indication
        appContent['timestamp'] = myDate.toJSON().replace('Z', '');
      }
      appContent['icone'] = payload.readUInt8(2);
      appContent['qaiRedDuration'] = { unit: 'min', value: payload.readUInt8(3) * 10 };
      appContent['temperatureMax'] = { unit: '\u00B0' + 'C', value: payload.readInt16BE(4) / 10 };
      appContent['co2Max'] = { unit: 'ppm', value: payload.readUInt16BE(6) };
      appContent['humidityMax'] = { unit: '%', value: payload.readUInt8(8) };
      appContent['vocMax'] = { unit: 'index value', value: payload.readUInt16BE(9) };
      // Content depends on network

      switch (payload.length) {
        case 11:
          break;
        case 25:
        case 29:
          appContent['temperatureMin'] = { unit: '\u00B0' + 'C', value: payload.readInt16BE(11) / 10 };
          appContent['co2Min'] = { unit: 'ppm', value: payload.readUInt16BE(13) };
          appContent['humidityMin'] = { unit: '%', value: payload.readUInt8(15) };
          appContent['vocMin'] = { unit: 'index value', value: payload.readUInt16BE(16) };
          appContent['temperatureAverage'] = { unit: '\u00B0' + 'C', value: payload.readInt16BE(18) / 10 };
          appContent['co2Average'] = { unit: 'ppm', value: payload.readUInt16BE(20) };
          appContent['humidityAverage'] = { unit: '%', value: payload.readUInt8(22) };
          appContent['vocAverage'] = { unit: 'index value', value: payload.readUInt16BE(23) };
          break;
        default:
          appContent.partialDecoding = PartialDecodingReason.MISSING_NETWORK;
          break;
      }
      return appContent;
    }
  }
}
