// @adeunis/codecs/src/products/comfort_serenity/0x6b.parser.ts
namespace codec {
  /**
   * Comfort Serenity 0x6b (alarm) frame parser
   */
  export class ComfortSerenity0x6bParser implements FrameParser {
    readonly deviceType = 'comfortSerenity';
    readonly frameCode = 0x6b;

    public parseFrame(payload: Buffer, configuration: Buffer, network: Network) {
      const appContent: Content = { type: '0x6b Comfort Serenity alarm' };
      const payloadLength = payload[1] & 0x04 ? payload.length - 4 : payload.length;

      if (payload[1] & 0x04) {
        // retrieval of the frame date timestamp value, then conversion to epoch 2013 (+1356998400), then in second
        const myDate = new Date((payload.readUInt32BE(payloadLength) + 1356998400) * 1000);
        // As we do not know if timestamp is locale or UTC , do remove 'UTC' indication
        appContent['timestamp'] = myDate.toJSON().replace('Z', '');
      }
      appContent['alarmTemperature'] = {
        alarmStatus: payload[2] & 0x10 ? 'active' : 'inactive',
        temperature: { unit: '\u00B0' + 'C', value: payload.readInt16BE(3) / 10 },
      };

      appContent['alarmHumidity'] = {
        alarmStatus: payload[2] & 0x01 ? 'active' : 'inactive',
        humidity: { unit: '%', value: payload.readUInt8(5) },
      };

      appContent['alarmCo2'] = {
        alarmStatus: payload[2] & 0x20 ? 'active' : 'inactive',
        co2: { unit: 'ppm', value: payload.readInt16BE(6) },
      };

      return appContent;
    }
  }
}
